### ISSUE TITLE
[BUG] Transductive search (EAS, ActiveSearch) never applies its gradient update

### ISSUE THREAD

**Issue by MoritzOster** [ai4co/rl4co#281]
## Describe the bug

`EAS` / `EASEmb` / `EASLay` and `ActiveSearch` compute a loss and backpropagate it, but never call `optimizer.step()`, so none of the parameters they are supposed to adapt is ever updated. Both methods currently reduce to plain (augmented, multi-start) sampling with incumbent re-injection.

There are two independent causes; either one alone breaks the update.

**1. `optimizer.step()` is missing.** In `EAS.training_step` (`rl4co/models/zoo/eas/search.py`) and `ActiveSearch.training_step` (`rl4co/models/zoo/active_search/search.py`):

```python
opt = self.optimizers()
opt.zero_grad()
self.manual_backward(loss)
# <- no opt.step()
```

The reference implementation ([`ahottung/EAS`](https://github.com/ahottung/EAS), `source/eas_emb.py`) does `optimizer.zero_grad()` / backward / `optimizer.step()`.

**2. `self.optimizers()` is not the optimizer holding the adapted parameters.** Both methods call `self.configure_optimizers(<subset of params>)` and discard the return value. `configure_optimizers` is a hook Lightning invokes once during `strategy.setup()`; calling it again by hand creates an optimizer that is never registered with the trainer. `self.optimizers()` therefore returns the optimizer built at setup time from `self.parameters()` — the whole policy, every parameter of which `EAS.on_train_batch_start` has just set to `requires_grad=False`.

So adding a bare `opt.step()` would step only frozen parameters whose `.grad` is `None`, and `opt.zero_grad()` never touches the adapted tensors, so their gradients accumulate across all search iterations instead of being reset.

## To Reproduce

`EASEmb` and `EASLay` adapt structurally different parameters, so with the same seed they must produce different results. They produce **bit-identical** results, which is only possible if neither of them adapts anything:

```python
import torch
from rl4co.envs import TSPEnv
from rl4co.models.zoo import EASEmb, EASLay
from rl4co.models.zoo.am import AttentionModelPolicy
from rl4co.utils.trainer import RL4COTrainer

results = {}
for cls in (EASEmb, EASLay):
    torch.manual_seed(0)
    env = TSPEnv(generator_params=dict(num_loc=20))
    policy = AttentionModelPolicy(env_name="tsp")
    model = cls(env, policy, env.dataset([4]), batch_size=2, max_iters=20, augment_size=8)
    RL4COTrainer(max_epochs=1, devices=1, logger=False, enable_checkpointing=False,
                 enable_progress_bar=False).fit(model)
    results[cls.__name__] = model.instance_rewards

print("EASEmb:", results["EASEmb"])
print("EASLay:", results["EASLay"])
print("identical:", torch.equal(results["EASEmb"], results["EASLay"]))
```

```bash
EASEmb: tensor([-5.4119, -6.5119, -5.8824, -6.1347], device='mps:0')
EASLay: tensor([-5.4119, -6.5119, -5.8824, -6.1347], device='mps:0')
identical: True
```

Instrumenting the tensor `EASEmb` adapts (`cached_embeds.logit_key`) confirms the mechanism directly — it has a non-zero gradient that is never zeroed, and it never moves:

```
adapted tensor: requires_grad=True  grad_norm=4.707e-01  max|delta|=0.000e+00
grad norm per iteration: 0.565 -> 0.844 -> 1.132 -> 1.299 -> 1.431 -> 1.635
self.optimizers() holds 82656 params, all with requires_grad=False
the optimizer built inside training_step holds 1280 params and is discarded
```

## Expected behavior

The parameters designated for adaptation (`cached_embeds.logit_key` for EAS-Emb, the `EASLayerNet` weights for EAS-Lay, all policy weights for `ActiveSearch`) should change during the search.

Note the rest of the EAS implementation matches the reference: all policy parameters are frozen; `EASEmb` adapts only `logit_key`, which is exactly POMO's `single_head_key` that official `eas_emb.py` optimizes; `EASLay` inserts the residual bottleneck on the MHA output before `project_out` with `W1`/`b1` Xavier- and `W2`/`b2` zero-initialised; and the loss is REINFORCE with a group-shared baseline over the sampled rollouts excluding the incumbent, plus `lambda * (-log p(incumbent))`. Only the optimizer step is missing.

## System info

- rl4co: reproduced on released `v0.6.0` and on `main` (`7a15396`); the files involved differ only cosmetically between the two
- torch 2.12.1, lightning 2.6.5, Python 3.13, macOS (also independent of accelerator/precision)

I have a fix ready and will open a PR referencing this issue.

*Disclaimer: both this report and the fix referenced above were co-authored with [Claude Code](https://claude.com/claude-code) (Claude Opus 5). The reproduction script and every measurement quoted above were reviewed and run locally before submitting.*
